cni-plugin/linkerd-cni-validator/src/main.rs

```diff
@@ -0,0 +1,169 @@
+        default_value = "linkerd_cni_validator=info,warn"
+    )]
+    log_level: kubert::LogFilter,
+
+    #[clap(long, default_value = "plain")]
+    log_format: kubert::LogFormat,
+
+    #[clap(long, default_value = "0.0.0.0:4140")]
+    outbound_proxy_addr: SocketAddr,
+
+    #[clap(parse(try_from_str), long, env = "KUBERNETES_SERVICE_HOST")]
+    target_ip: IpAddr,
+
+    #[clap(parse(try_from_str), long, env = "KUBERNETES_SERVICE_PORT")]
+    target_port: u16,
+}
+
+static REDIRECT_RESPONSE: &str = "REDIRECTION SUCCESSFUL";
+
+// ERRNO 95: Operation not supported
+const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
+    let Args {
+        log_level,
+        log_format,
+        outbound_proxy_addr,
+        target_ip,
+        target_port,
+    } = Args::parse();
+
+    log_format.try_init(log_level)?;
+    let target_addr = SocketAddr::new(target_ip, target_port);
+
+    info!(%outbound_proxy_addr, %target_addr, "Validating outbound traffic is redirected to proxy's outbound port");
+
+    let (shutdown_tx, shutdown_rx) = kubert::shutdown::sigint_or_sigterm()?;
+    let (ready_tx, ready_rx) = tokio::sync::oneshot::channel::<Result<()>>();
+
+    let server_task = tokio::spawn(outbound_serve(outbound_proxy_addr, ready_tx, shutdown_rx));
+
+    let validation_task = tokio::spawn(validate_outbound_redirect(target_addr, ready_rx));
+
+    tokio::select! {
+        task = validation_task => {
+            let validation_result = task.expect("Failed to run validator task");
+            if let Err(err) = validation_result {
+                error!(error = %err, "Failed validation");
+                std::process::exit(UNSUCCESSFUL_EXIT_CODE);
+            }
+
+            info!("Validation passed successfully...exiting");
+            Ok(())
+        }
+
+        _ = shutdown_tx.signaled() => {
+            if let Err(e) = server_task.await.expect("Failed to run outbound server task") {
+                error!(error = %e, "Failed to validate outbound routing configuration");
+                std::process::exit(UNSUCCESSFUL_EXIT_CODE);
+            }
+
+            error!("Failed to validate due to server terminating early");
+            std::process::exit(UNSUCCESSFUL_EXIT_CODE);
+        }
+    }
+}
+
+async fn validate_outbound_redirect(
+    target_addr: SocketAddr,
+    ready_rx: tokio::sync::oneshot::Receiver<Result<()>>,
+) -> Result<()> {
+    let timeout = std::time::Duration::from_secs(120);
+    if (tokio::time::timeout(timeout, ready_rx).await).is_err() {
+        anyhow::bail!("timed-out ({:?}) waiting for server to be ready", timeout);
+    }
+
+    let mut stream = {
+        debug!(original_dst = %target_addr, "Building validation client");
+        let socket = TcpStream::connect(target_addr).await?;
+        let peer = socket.peer_addr()?;
+        debug!(original_dst = %peer, "Client connected to validation server");
```

Comment:
`target_addr` and `peer_addr` should always be the same here, right? could we maybe replace this with
```suggestion
        debug!(original_dst = %target_addr, "Building validation client");
        let socket = TcpStream::connect(target_addr).await?;
        assert_eq!(target_addr, socket.peer_addr().unwrap());
        debug!(original_dst = %target_addr, "Client connected to validation server");
```